### PR TITLE
Implement kubeadm bootstrap token management

### PR DESCRIPTION
**Pull request by dgoodwin**
Creates bootstrap tokens as secrets per the specification in #30707 

_WARNING_: These are not currently hooked up to the discovery service or the token it creates.

Still TODO:
- [x] delete tokens
- [x] merge with #35144 and adopt it's testing approach
- [x] determine if we want wholesale json output & templating like kubectl (we do not have an API object with the data we want here) may require a bit of plumbing.
- [x] allow specifying a token duration on the CLI
- [x] allow configuring the default token duration
- [x] hook up the initial token created during init

Sample output:

```
(root@centos1 ~) $ kubeadm token create
Running pre-flight checks
<cmd/token> Token secret created: f6dc69.c43e491752c4a0fd
(root@centos1 ~) $ kubeadm token create
Running pre-flight checks
<cmd/token> Token secret created: 8fad2f.e7b78c8a5f7c7b9a
(root@centos1 ~) $ kubeadm token list  
Running pre-flight checks
ID        TOKEN                     EXPIRATION
44d805    44d805.a4e78b6cf6435e33   23h
4f65bb    4f65bb.d006a3c7a0e428c9   23h
6a086e    6a086e.2ff99f0823236b5b   23h
8fad2f    8fad2f.e7b78c8a5f7c7b9a   23h
f6dc69    f6dc69.c43e491752c4a0fd   23h
f81653    f81653.9ab82a2926c7e985   23h
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35805)

**Release note**:
```release-note
Added kubeadm commands to manage bootstrap tokens and the duration they are valid for.
```

### CODE REVIEW

**Comment by dgoodwin** [comment]
CC @jbeda @kubernetes/sig-cluster-lifecycle

**Comment by luxas** [comment]
@dgoodwin Really nice! This is important
Is there anything I can do to help? Reviewing, etc..?

I understand that this is wip, so I'm asking before I touch it :)

File: cmd/kubeadm/app/master/apiclient.go
Comment (pires) [inline cmd/kubeadm/app/master/apiclient.go]: Empty line.

File: cmd/kubeadm/app/tokens/tokens.go
Comment (pires) [inline cmd/kubeadm/app/tokens/tokens.go]: Please, remove empty import.

**Review by pires — CHANGES_REQUESTED** [review]
Few nits. Also no tests or is it in-flight?

**Comment by luxas** [comment]
@pires This is in my understanding a really rough first draft; see the TODO-list in the description

**Comment by pires** [comment]
@luxas you're right, I should've done that instead of jumping to a code review.

@dgoodwin @jbeda probably should split this PR into several smaller ones? It would make it easier to review/merge. You could create an umbrella issue for keeping track of what's been delivered (as in merged).

**Comment by dgoodwin** [comment]
@pires splitting the overall discovery implementation up into multiple PRs makes sense to me, this one is quite independent of the discovery API and auth work, hopefully we can push those all in separately. If you meant splitting *this* pr up let me know what you have in mind, it might slow things down as they would need sequential merges. (or if you meant merging this soon then more PRs for todos that might work fine)

@luxas @pires what do you guys think about the -o json and templating support? It involves a bit of wiring that looked a little nasty, but we could try if you feel it's important. It's a little strange here because we don't have an API type for this, the secret itself is unusable for the user, it has to be interpreted and displayed differently. I'm not sure if we can just make a type for that purpose and get away with it, most likely we could. (giving it a fake "kind" perhaps) All depends if folks think it's important though, we don't do that anywhere else in kubeadm and maybe we shouldn't bother.

**Comment by pires** [comment]
> If you meant splitting this pr up let me know what you have in mind, it might slow things down as they would need sequential merges. (or if you meant merging this soon then more PRs for todos that might work fine)

Yeah, I meant to have those PRs merged quickly, so they need to be easy (fast) to review and merge. If it's a big PR, I'm afraid it may become a burden for you (rebase hell) and the reviewers.

Re: templating, I don't have an opinion, sorry. However, AFAIK `kubeadm` has its own API objects, alright (`MasterConfiguration`, `NodeConfiguration` and `ClusterInfo`).

File: cmd/kubeadm/app/cmd/token.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/token.go]: I guess we should have an option to specify the token that should be created.
I think something like this is fine
```
# kubeadm create token abcdef.1234567890123456 --duration 8h
abcdef.1234567890123456
# kubeadm create token --duration 8h
fedcba.6543210987654321
```

File: cmd/kubeadm/app/cmd/token.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/token.go]: Can this be a duration instead of a string?

File: cmd/kubeadm/app/cmd/token.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/token.go]: Don't we have some code like this already in `apiclient.go`? If not, we should refactor out this code into a helper

File: cmd/kubeadm/app/cmd/token.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/token.go]: I think we should just print the token here

**Review by luxas — COMMENTED** [review]
Quick pass, some design comments

File: cmd/kubeadm/app/util/tokens.go
Comment (luxas) [inline cmd/kubeadm/app/util/tokens.go]: can you fix this at the same time :)?

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: Wow it can, cool.

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: Right now it's in as kubeadm token create --token = MY.TOKEN. Let me know if you'd prefer the argless route.

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: It was based on that but quite different, however on a second pass I was able to merge the two and this is now a shared method.

File: cmd/kubeadm/app/util/tokens.go
Comment (dgoodwin) [inline cmd/kubeadm/app/util/tokens.go]: Of course, done!

File: cmd/kubeadm/app/cmd/token.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/token.go]: I think I prefer the argless route, but we should defer that decision until you've proceeded more on this PR and until others are taking a look at it

File: cmd/kubeadm/app/cmd/token.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: We should make this an `upsert` where we update or create/insert.

Do this:
1. Run in a loop with a certain number of retries in case something flakes out.  Something like 5.
2. Try to get the secret.  
  1. If it is there, then replace the Data array and update.
  2. If it doesn't exist (`IsNotFound`) then create it and set it
3. If you get an error then retry from the top.  Errors might be `IsNotFound` (if you are doing an update and it was deleted between the get/update) or `IsConflict` (if you raced someone to do an update and lost).


See `pkg/api/errors/errors.go`

File: cmd/kubeadm/app/cmd/token.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: `data["usage-bootstrap-signing"] = []byte("true")`

File: cmd/kubeadm/app/cmd/token.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: Should this go to "out"?

File: cmd/kubeadm/app/cmd/token.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: Eventually folks are going to ask for ways to script this and we'll find ourselves reimplementing all of the output modes that `kubectl` has.  This is good for now but won't last.

Also -- most users see the key as `123456.1234567890123456`.  I'm not sure we need to break it down to id and token (I called that part "secret").  I'd just name the column "token" and have the dotted form.

File: cmd/kubeadm/app/cmd/token.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: Handle the case where this doesn't exist and it doesn't expire.

File: cmd/kubeadm/app/cmd/token.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: there is no guarantee that these secrets are well formed.  Use the tuple return form (`tokenId, ok := secret.Data["token-id"]` and check `ok`.

If keys are malformed, I'd output a warning but continue on.  There will be some good keys that folks can use.   That means you probably want to send stderr in too.

File: cmd/kubeadm/app/cmd/token.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: You may get `IsNotFound` here and that is okay.  That means that it was deleted while the user was doing something else -- perhaps expired?  I'd output something like "<cmd/token> bootstrap token already deleted: <tokenId>".  This shouldn't return an error.

File: cmd/kubeadm/app/master/apiclient.go
Comment (jbeda) [inline cmd/kubeadm/app/master/apiclient.go]: This isn't a token thing, necessarily.  This class will get used lots of places.

We should either pass the "phase" in or prepend the phase as we dump the error.

**Review by jbeda — CHANGES_REQUESTED** [review]

**Comment by jbeda** [comment]
Very cool!  No huge changes requested.  Biggest thing is around API server errors.

Do we want a release note here?  If we are releasing kubeadm with the rest of kubernetes we probably do.

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: TTL sounds good. 

Mentioning no TTL feels a little worrisome, do you feel that's important to expose if users want to opt into it? Or could we get away with just not offering an unlimited bootstrap token.

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: Seeing below I may have misunderstood.

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: I implemented delete via the token ID, so my goal was to give the user two easy to copy parts, ID and full token which contains both parts with the '.'.

Are you suggesting just columns "token" and "expiration"? If so I could perhaps have "delete" handle both forms, and automatically dissect it if given the whole token.

File: cmd/kubeadm/app/cmd/token.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: I'm cool with leaving "0 == unlimited" undocumented.  It is bad practice but I could see folks wanting to do it for test clusters.  We should make sure we can display unlimited tokens though in "list".

File: cmd/kubeadm/app/cmd/token.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: Ah -- gotcha.  So the columns are `ID`, `ID.Secret`, `TTL`.  I misread it before and thought the second column was just the secret.  This looks good to me.

**Comment by jbeda** [comment]
@dgoodwin -- This is really shaping up.  But we are going to want to be careful how we stage this.  Specifically, we shouldn't make this the default for users until we get the server side stuff done and these tokens work for both discovery and TLS and are the only thing we use.  Otherwise we might just confuse users.

Perhaps put this under a "devel" or "experimental" node in the command structure so that for now it is `kubeadm devel token list`?

File: cmd/kubeadm/app/master/apiclient.go
Comment (dgoodwin) [inline cmd/kubeadm/app/master/apiclient.go]: These were intended to just be debug info identifying what module/file the message was from, IIRC they might actually be short lived and going away in future. This one was a bad copy paste, it should be master/apiclient still, propose I fix it to master/apiclient and we'll sort that out with output cleanup in the future. It the method gets moved to a more generic module we can update prefix then as well.

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: Sounds good, done.

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: Done!

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: Added.

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: This should now be corrected.

File: cmd/kubeadm/app/cmd/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/token.go]: Agree, done!

**Comment by dgoodwin** [comment]
We use "ex" in some openshift commands for experimental, how does that sound? (will add it this way for now as it's nice and snappy, if anyone feels strongly it should be elsewhere let me know)

**Comment by dgoodwin** [comment]
I think I've addressed everything mentioned above. Separate commits for re-reviewability.

File: cmd/kubeadm/app/util/tokens.go
Comment (jbeda) [inline cmd/kubeadm/app/util/tokens.go]: The only "hole" here is if someone creates a bootstrap token from some other client with a different name but the same ID.  The "right" way to do this would be to list all the tokens and index those by ID.  Then you could update a token without making assumptions about the secret name.

But that probably isn't worth the trouble at this point.  As things stand the server impl will throw a warning and do its best.  Folks can always use a `kubectl delete secret` command to clean this up if need be.

**Review by jbeda — APPROVED** [review]

**Comment by jbeda** [comment]
Looking great!  Squash and I'll LGTM.

**Comment by jbeda** [comment]
Also, the bazel verification is failing...

**Comment by dgoodwin** [comment]
I rebased on top of @pipejakob 's work for the generate command, I kept it as a separate command, I briefly considered a create --print-only option but then we need to start watching for combinations with --ttl and --token, seemed best to just keep it separate for now.

@pipejakob please take a quick look at let me know if I botched anything in the merge.

Squashed and signed otherwise, should be ready to go.

**Comment by dgoodwin** [comment]
Fixed another verify failure.

**Comment by jbeda** [comment]
It looks good -- let me know when you want me to LGTM.

**Comment by dgoodwin** [comment]
Just caught an issue this morning where the token command was also available top level as well as below "ex", removed now.

**Comment by dgoodwin** [comment]
Ready when you are @jbeda

**Comment by luxas** [comment]
@k8s-bot unit test this

**Comment by luxas** [comment]
@k8s-bot cvm gke e2e test this

**Comment by luxas** [comment]
LGTM, also it looked good to @jbeda earlier.

**Comment by luxas** [comment]
@dgoodwin This unit failure looks real. PTAL

**Comment by dgoodwin** [comment]
Thanks @luxas, updated the tests to "kubeadm ex token".

**Comment by luxas** [comment]
@dgoodwin You can now rebase this and we'll get it merged :+1:

**Comment by dgoodwin** [comment]
@k8s-bot test this

**Comment by dgoodwin** [comment]
@k8s-bot test this

**Comment by dgoodwin** [comment]
@luxas waiting on builds but I think this is ready now, big rebase conflict yesterday and then got bazel fixed this morning.

**Comment by dgoodwin** [comment]
@k8s-bot cvm gke e2e test this

**Comment by luxas** [comment]
@jbeda @mikedanese PTAL once more

**Review by luxas — APPROVED** [review]
Haven't had a chance to manually test this yet unfortunately, but it looks good to me.

I don not want to defer this PR anymore, let's just get it in and move on! We'll have lots of improvement possibilities to actually hook this flow up with something once @jbeda's PR is merged.

@jbeda @mikedanese Can I apply the label?
Does the rebase stuff look good?

**Comment by jbeda** [comment]
It LGTM but I think the only open question is `beta` in the secret type.   @mikedanese -- thoughts there?

**Comment by luxas** [comment]
@jbeda You can update it in your PR anyway, since you will have to rebase yours in any case :smile:

**Comment by jbeda** [comment]
/lgtm

Final comment -- we should probably have a release note on this.  Can you add one to the description and reset the labels (or ping me to tweak them).

**Comment by dgoodwin** [comment]
@jbeda I added a release note block in the description, I don't think I can change labels myself however. Let me know if there's anything else I need to do, thanks!

**Comment by jbeda** [comment]
Fixed.  Thanks!

**Comment by luxas** [comment]
@jbeda Just a side note, I'm absolutely not opposed to adding release-note labels for kubeadm PRs, but we have not done it with any PRs so far I think, at least not many.

I've aggregated the relnotes manually instead so far.
I'm not opposed to starting to add release-note labels for kubeadm now in time for v1.6, but since we're gonna break out soon or at least in a reasonably short timeframe, I'm not sure it's worth it to start a pattern that we know we're gonna break

**Comment by luxas** [comment]
Hahah, seems like I repeated myself there a bit (writing half a comment - doing something else - writing the other part - publish :smile:)

**Pull request merged**
